pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java

```diff
@@ -0,0 +1,250 @@
+ */
+
+package net.sourceforge.pmd.lang.java.rule.bestpractices;
+
+import static net.sourceforge.pmd.properties.PropertyFactory.enumProperty;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import net.sourceforge.pmd.lang.ast.AbstractNode;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.java.ast.ASTAssignmentOperator;
+import net.sourceforge.pmd.lang.java.ast.ASTBlockStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTForInit;
+import net.sourceforge.pmd.lang.java.ast.ASTForStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTLocalVariableDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTName;
+import net.sourceforge.pmd.lang.java.ast.ASTPostfixExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTPreDecrementExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTPreIncrementExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclaratorId;
+import net.sourceforge.pmd.lang.java.rule.performance.AbstractOptimizationRule;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+
+public class AvoidReassigningLoopVariablesRule extends AbstractOptimizationRule {
+
+    private static final Map<String, ForeachReassignOption> FOREACH_REASSIGN_VALUES;
+
+    static {
+        final Map<String, ForeachReassignOption> map = new HashMap<>();
+        map.put("deny", ForeachReassignOption.DENY);
+        map.put("firstOnly", ForeachReassignOption.FIRST_ONLY);
+        map.put("allow", ForeachReassignOption.ALLOW);
+        FOREACH_REASSIGN_VALUES = Collections.unmodifiableMap(map);
+    }
+
+    private static final PropertyDescriptor<ForeachReassignOption> FOREACH_REASSIGN
+            = enumProperty("foreachReassign", FOREACH_REASSIGN_VALUES)
+            .defaultValue(ForeachReassignOption.DENY)
+            .desc("how/if foreach control variables may be reassigned")
+            .build();
+
+    private static final Map<String, ForReassignOption> FOR_REASSIGN_VALUES;
+
+    static {
+        final Map<String, ForReassignOption> map = new HashMap<>();
+        map.put("deny", ForReassignOption.DENY);
+        map.put("skip", ForReassignOption.SKIP);
+        map.put("allow", ForReassignOption.ALLOW);
+        FOR_REASSIGN_VALUES = Collections.unmodifiableMap(map);
+    }
+
+    private static final PropertyDescriptor<ForReassignOption> FOR_REASSIGN
+            = enumProperty("forReassign", FOR_REASSIGN_VALUES)
+            .defaultValue(ForReassignOption.DENY)
+            .desc("how/if for control variables may be reassigned")
+            .build();
+
+    public AvoidReassigningLoopVariablesRule() {
+        definePropertyDescriptor(FOREACH_REASSIGN);
+        definePropertyDescriptor(FOR_REASSIGN);
+    }
```

Comment:
Since your Rule doesn't store state between visits of different AST nodes (in fact you're in the special case where you only visit one node type), you can use the rule chain to get a big performance boost. The only thing to do here is to add a call to `addRuleChainVisit(ASTLocalVariableDeclaration.class);` in your constructor

Reply:
Nice, I didn't know about that... Done!